src/api.ts

```diff
@@ -792,17 +800,40 @@ function createApi({
         style: styleToSave,
       })
     },
-
     async listStyles() {
+      // `bytesStored` calculates the total bytes stored by tiles that the style references
+      // Eventually we want to get storage taken up by other resources like sprites and glyphs
       return db
         .prepare(
-          "SELECT Style.id, json_extract(stylejson, '$.name') as name FROM Style"
+          `
+          SELECT Style.id, 
+            json_extract(Style.stylejson, '$.name') as name, 
+            (
+              SELECT SUM(LENGTH(TileData.data))
+              FROM TileData
+              JOIN Tile ON Tile.tileHash = TileData.tileHash
+              JOIN Tileset ON Tileset.id = Tile.tilesetId
+              JOIN (
+                SELECT S2.id AS styleId, StyleReferencedTileset.value AS tilesetId
+                FROM Style S2, json_each(S2.sourceIdToTilesetId, '$') AS StyleReferencedTileset
+              ) AS SourceTileset ON SourceTileset.tilesetId = Tileset.id
+              JOIN Style S1 ON S1.id = SourceTileset.styleId
+              WHERE SourceTileset.styleId = Style.id
+            ) AS bytesStored
+          FROM Style;
```

Comment:
Complicated query! I can't quite parse it, but if it works, it works. It would be really good to collect metrics on this (running on phones) to keep track if this is a performance bottle neck. Let's leave for now since it seems sqlite runs really fast.

Reply:
another thing to note is that this query has lots of joins because I'm trying to adhere to base db principles of using relationships, which means using the Tile table (which is a joins table between TileData and Tileset). Theoretically could be made simpler by using the tilesetId field in the TileData table, but I have reservations about the existence of that column. Can't fully remember the reason for adding it (maybe due to a previous conversation about reducing the need for the joins table? or maybe making updates/inserts less complicated in general?), but I'd advocate for removing it if the reason isn't entirely clear - just to avoid potential issues with normalization  or tech debt

Reply:
The reason for having the `tilesetId` field on the tileData table is to manage deletions and updates of a tileset. Without it, it is hard (impossible?) to know what can be updated / deleted. Imagine a 2 tilesets each of which have the same colour of blue tile (e.g. for ocean, or river), such that the hash for the tiles is the same. Now imagine the user updates one of those tilesets to have green rivers / ocean. It would be a costly query to check all the references to know whether the blue tile is still used by another tileset. Because tilesets can be hundreds of thousands of tiles, my thought was tracing the back-references for each of those tiles for any deletion query is just too costly, hence putting the tilesetId in the tiledata table. I think the same reasoning applies for this query.
Would it not work in this query to join the tileData directly using the StyleReferencedTileset?

Reply:
i now remember you explaining this previously. it makes more sense now but makes me wonder if there's still utility in having the joins table in that case (again, you may have answered this question before already). does keeping the joins table help with reading data whereas the field in question helps with mutating it? just feels like there's some kind of redundancy because the ultimate goal for each is to associate tiles with tilesets

> Would it not work in this query to join the tileData directly using the StyleReferencedTileset?

i think it potentially could! although a quick test using the sample db i have produced different results 🤔 

1. Using `Tile` table:
    - Query:

       ```sql
       SELECT Style.id, 
       json_extract(Style.stylejson, '$.name') as name, 
       (
         SELECT SUM(LENGTH(TileData.data))
         FROM TileData
         JOIN Tile ON Tile.tileHash = TileData.tileHash
         JOIN Tileset ON Tileset.id = Tile.tilesetId
         JOIN (
           SELECT S2.id AS styleId, StyleReferencedTileset.value AS tilesetId
           FROM Style S2, json_each(S2.sourceIdToTilesetId, '$') AS StyleReferencedTileset
         ) AS SourceTileset ON SourceTileset.tilesetId = Tileset.id
         JOIN Style S1 ON S1.id = SourceTileset.styleId
         WHERE SourceTileset.styleId = Style.id
       ) AS bytesStored
       FROM Style;
       ```
   - Result:
 
      | name | bytesStored |
      |-------|-----------|
      | tiles | 1303505 |
      | Countries | **7970977** |


2. Without using `Tile` table:
    - Query:

       ```sql
       SELECT Style.id, 
       json_extract(Style.stylejson, '$.name') as name, 
       (
         SELECT SUM(LENGTH(TileData.data))
         FROM TileData
         JOIN (
           SELECT S2.id AS styleId, StyleReferencedTileset.value AS tilesetId
           FROM Style S2, json_each(S2.sourceIdToTilesetId, '$') AS StyleReferencedTileset
         ) AS SourceTileset ON SourceTileset.tilesetId = TileData.tilesetId
         JOIN Style S1 ON S1.id = SourceTileset.styleId
         WHERE SourceTileset.styleId = Style.id
       ) AS bytesStored
       FROM Style;
       ```
   - Result:
 
      | name | bytesStored |
      |-------|-----------|
      | tiles | 1303505 |
      | Countries | **7423783** |
      
      
Will take a minute to figure out which one is actually more accurate. Will edit this comment when I have that calculated

Reply:
It would be good to do row counts on each of these. If there are duplicate tiles (e.g. tiles with the same hash) then the first query might be counting the same row in the tileData table more than once.

Reply:
Checked the original [mbtiles file](https://github.com/digidem/mapeo-map-server/blob/2dbdc55e7618630e377d3c7995b2c04d2a71dcbf/src/fixtures/mbtiles/raster/countries-png.mbtiles) used for the `Countries` tileset and ran this query to check the byte count:

```sql
SELECT SUM(LENGTH(tile_data)) FROM tiles;
```

and it returned `7970977`, which aligns with the first query. Either the second query is missing something or it just isn't equivalent?

Reply:
Ah yeah i'm starting to better understand what's going on. It's like you mentioned, where query 1 is probably counting rows twice due to repeating tile hashes. I see in this query that some tiles are deduped because they have the same tile hash. Very cool!

```sql
SELECT TileData.tileHash, COUNT(*) as tileCount
FROM TileData
JOIN Tile ON Tile.tileHash = TileData.tileHash
JOIN Tileset ON Tileset.id = Tile.tilesetId
GROUP BY TileData.tileHash
ORDER BY tileCount DESC;
```
<img width="373" alt="image" src="https://user-images.githubusercontent.com/18542095/176480818-43223f4e-cd1e-40b0-b9ad-b02574b858bd.png">

Reply:
alright so my conclusion for all of this is that query 2 is what we want to use in this case, since we're interested in the true amount of tile data that's stored on disk. what query 1 really does is expose how many bytes are exposed for a style's tilesets when actually being used, but that includes potential duplicated use of tiles.

Reply:
That's what the query is doing since the `tiles` table in mbtiles holds the tile data (i.e. the actual image bytes)